### ISSUE TITLE
INFO[0101] failed to verify token: token signed by untrusted key with ID:

### ISSUE THREAD

**Issue by zfLQ2qx2** [distribution/distribution#4299]
### Description

I'm testing updating to a recent build of docker-registry but I'm getting stuck on the authorization part. The error in the file is:

INFO[0101] failed to verify token: token signed by untrusted key with ID: "xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx"

- I've generated a completely new self-signed certificates w/ new a RSA 2048 key
- I've verified the key and certificate modulus matches
- I've verified that both the registry and auth service have the same key and cert, with matching modulus
- I've verified that the auth service is returning a positive response with a JWT token signed by the right key

The auth service is using docker/libtrust to generate the keyid:
https://github.com/cesanta/docker_auth/blob/38e7252690dc31ab7ccd4185a2dead973099f37c/auth_server/server/server.go#L384
https://github.com/cesanta/docker_auth/blob/38e7252690dc31ab7ccd4185a2dead973099f37c/auth_server/server/config.go#L88

So it seems to be docker-registry itself making the determination that the key is untrusted. However it doesn't generate any errors or warnings when loading the key, only when processing the auth response. The error message doesn't provide any hints as to why it is untrusted, and there are no other relevant messages in the log. This last worked using a build of docker-registry from early 2022 (commit c202b9b0d7b79a67337dec8e1f1bafb1c7095315), the only change is using a March 2024 build (commit 663b430ccc10e184ecb82eff77661a38f1f6f660) - though I also tried a newer certificate to see if it gave a different result. 

### Reproduce

Setting up a local registry can be complex, I'll wait to see if its something simple like docker-registry only supports rsa-4096, or has banned values for common name, etc.

### Expected behavior

The server should accept the positive response from the auth server and allow access.

### registry version

/usr/bin/docker-registry github.com/distribution/distribution/v3 v3.0.0-alpha.1.m+unknown

**Comment by zfLQ2qx2** [comment]
Hmm.. it looks like a workaround might be to have the auth service not send a key id, or to have the registry not validate it, but is that a good solution?

[token signed by untrusted key with ID](https://github.com/distribution/distribution/blob/main/registry/auth/token/token.go#L146-L150)

**Comment by milosgajdos** [comment]
I feel like your trusted keys are not being loaded properly or you're not loading them at all:
https://github.com/distribution/distribution/blob/d9815da9cba94fd9ca2e52c4781f0373e57323a8/registry/auth/token/accesscontroller.go#L282-L287

You need to provide a path to your [JWKS](https://datatracker.ietf.org/doc/html/rfc7517#section-5) and pass it to the config file via the `jwks` config:
https://github.com/distribution/distribution/blob/d9815da9cba94fd9ca2e52c4781f0373e57323a8/registry/auth/token/accesscontroller.go#L265-L270

You can of course "just" rely on the Cert bundle path to which is also passed via a config param:
https://github.com/distribution/distribution/blob/d9815da9cba94fd9ca2e52c4781f0373e57323a8/registry/auth/token/accesscontroller.go#L258-L263

**Comment by zfLQ2qx2** [comment]
@milosgajdos I tried that and if the key is not present in the registry config it results in a panic - "panic: unable to configure authorization (token): token auth requires at least one token signing key". Likewise on the auth server I can see that the KeyID changes if I update the key but remains stable if I don't, so I'm pretty confident it is loading the file. I tried not sending the keyID and that resulted in the message "INFO[0050] failed to verify token: go-jose/go-jose: no x5c header present in message".

I am wondering if the format of the KeyID field changed and docker-registry can no longer handle values produced by a different method.

**Comment by zfLQ2qx2** [comment]
This seems to be the breaking change: https://github.com/distribution/distribution/pull/4096

Docker-registry can no longer process the kid values that were produced by the docker-libtrust library. It's not clear if its an issue of making the authentication service produce kid values the registry does understand, adding some compatibility code to honor the format that docker-libtrust used (the presence of colons seems to to be an indicator that its a docker-libtrust style id, or if there are other changes required such as the x5c field referenced above.

**Comment by zfLQ2qx2** [comment]
Confirmed that if I run https://github.com/milosgajdos/distribution/tree/1d410148efe6d1b7fd56457507a9dd465b105ec4 instead, authentication works again.

**Comment by milosgajdos** [comment]
> if the key is not present in the registry config it results in a panic - "panic: unable to configure authorization (token): token auth requires at least one token signing key"

This is the correct behaviour -- how do you want to verify the token if you don't provide the trusted keys to registry?

> I am wondering if the format of the KeyID field changed and docker-registry can no longer handle values produced by a different method

FYI: we have deprecated `libstrust` because it's no longer active and in fact, it has been archived by Docker. But I digress. 

The format of the `keyID` is irrelevant here as long as the id is unique within the JWKs `keys` field. In my original answer I've linked to the JWKs spec: https://datatracker.ietf.org/doc/html/rfc7517#section-5

auth0 have a nice breakdown of what the JWKs look like: https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-set-properties

We load the JWKs we find in the file specified via the config: https://github.com/distribution/distribution/blob/d9815da9cba94fd9ca2e52c4781f0373e57323a8/registry/auth/token/accesscontroller.go#L282-L287

> I tried not sending the keyID and that resulted in the message "INFO[0050] failed to verify token: go-jose/go-jose: no x5c header present in message".

If you don't send the keyID and load the JWKs then the verification should correctly fail - how is the registry to know what key to verify if it doesn't have the keyID to look it up in its trusted keys? The error you are seeing happens when you rely on TLS certs but again fail to get registry to load them so it can use them for verification.

**Comment by markuskirch** [comment]
@milosgajdos we're encountering the same problem while testing 3.0.0-alpha.1, although our compose configuration looks fine to me.

What am I missing here?

compose configuration of the registry:
`  registry:
    image: registry:3.0.0-alpha.1
    restart: unless-stopped
    environment:
      REGISTRY_STORAGE_DELETE_ENABLED: true
      REGISTRY_AUTH: token
      REGISTRY_AUTH_TOKEN_REALM: "https://<<auth-server-domain>>/auth"
      REGISTRY_AUTH_TOKEN_SERVICE: "Docker registry"
      REGISTRY_AUTH_TOKEN_ISSUER: "MyOrg"
      REGISTRY_AUTH_TOKEN_ROOTCERTBUNDLE: "/cert/certificate.pem"
    volumes:
      - docker-registry:/var/lib/registry
      - token-certificate:/cert:ro`
      
the volume mounted at `/cert` contains exactly two files: certificate.pem and private.key and is mounted both on the registry and the auth server containers. The auth server generates tokens based on the same certificate.pem and private.key.

If I call the auth server directly e.g., `curl -u "user:password" "https://<<auth-server-domain>>/auth"`, authentication is successful (a token is returned)
If I call `docker login <<registry-domain>>:5000` and enter user:password, I get a 'failed with status: 401 Unauthorized' response and the registry logs the error "failed to verify token: token signed by untrusted key with ID: XXX"

Thanks for your help!

**Comment by milosgajdos** [comment]
Can you please format your message properly. It's impossible to parse useful info from it, I'm afraid.

**Comment by markuskirch** [comment]
@milosgajdos I'm very sorry for my cryptic comment. I just updated it with a detailed explanation of the error and a sample configuration to reproduce it locally.

Thanks for your great work maintaining this essential project!
Please let me know if there's anything I can do to help

**Comment by milosgajdos** [comment]
Thanks, that's useful. Off the top of my head, I can see that your JWT contains the [_optional_ `kid` value.](https://www.rfc-editor.org/rfc/rfc7515#section-4.1.4)

>  The "kid" (key ID) Header Parameter is a hint indicating which key
>  was used to secure the JWS.  This parameter allows originators to
>  explicitly signal a change of key to recipients.  The structure of
>  the "kid" value is unspecified.  Its value MUST be a case-sensitive
>  string.  Use of this Header Parameter is OPTIONAL.
>
>  When used with a JWK, the "kid" value is used to match a JWK "kid"
>  parameter value.

If you are going to include the `kid`, then you need to tell the registry that you trust the keys that sign that were used to sign it so we can verify it. That's what trusted keys config is for. You can read about JWKs here:  https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-set-properties

Are you signing these keys with TLS certs? 

Now I see where all these woes are coming from...The original code was checking the certs chain first and if that passed the validation, we'd just return; the current code checks the JWTs first and certs later, which surfaced a lot of issues to folks.

I've opened a [PR in `go-jose/go-jose`](https://github.com/go-jose/go-jose/pull/117) that exposes the missing `x5c` header in JWT that would help us revert to the original logic, but beware of what you're trying to do in your code now; if you want to make things work as is now, you need to provide trusted keys to registry so registry can verify the kid in your JWT.

**Comment by markuskirch** [comment]
Thanks for your quick help!

I noticed earlier today that `kid` is an optional parameter in the JWS spec. Interesting.
The third-party token service we currently use is becoming less and less maintained, which doesn't make me feel too optimistic about opening a PR there. Seems like I may have to sit down and write my own after all...

Please excuse my foolish question: What exactly is the "trusted keys" config and where can I specify the registry to trust a certain key?
I found nothing promising around the auth section in the [list of configuration examples](https://distribution.github.io/distribution/about/configuration/#overriding-the-entire-configuration-file).

We indeed use TLS certificates to generate tokens. In production, the registry is deployed behind a load-balancer that handles TLS for all services. Therefore, no `http.tls.[]` parameters are set in our registry configuration file.

**Comment by milosgajdos** [comment]
Yeah we're missing `jwks` section in the docs:

https://github.com/distribution/distribution/blob/c49220d492233615f22c1eea91e3121a681751eb/registry/auth/token/accesscontroller.go#L184

Basically, it lets you specify the path to the keys you can use for verifying the JWT tokens -- in the similar way as the cert bundle path is specified.

If you are using TLS certs to sign your tokens, then once my PR is merged into `go-jose` I'll open a PR to this repo and we'll switch the order of verification; shoehorning it without proper error handling would just add more tech debt

**Comment by milosgajdos** [comment]
I've opened a draft PR with potential fix https://github.com/distribution/distribution/pull/4349

Feel free to verify

**Comment by milosgajdos** [comment]
@zfLQ2qx2 mind checking if this is still an issue? See https://github.com/distribution/distribution/pull/4349

**Comment by milosgajdos** [comment]
Closing, addressed by https://github.com/distribution/distribution/pull/4349. Feel free to reopen.

**Comment by zhangyoufu** [comment]
@milosgajdos There is still a logic error in the implementation of VerifySigningKey, introduced in #4349.

https://github.com/distribution/distribution/blob/21f3291612301af9c5c9e6a839504ea140844b05/registry/auth/token/token.go#L175-L189

When `verifyCertChain` returns `jose.ErrMissingX5cHeader`, we head to the `switch` statement. In the 2nd case, we found a matching keyid in `verifyOpts.TrustedKeys`, the `err` variable is not cleared before returning.

**Comment by milosgajdos** [comment]
Can you open a PR so we can discuss this.

**Comment by milosgajdos** [comment]
@zhangyoufu thanks for highlighting the issue in the token verification code!

I've reopened the issue and also opened a PR that fixes the logic error: https://github.com/distribution/distribution/pull/4415

PTAL

**Comment by zhangyoufu** [comment]
> I've reopened the issue and also opened a PR that fixes the logic error: #4415
> 
> PTAL

LGTM.

---

I ran into this issue during migration from v2.8 to v3.0. And I believe many many more will. 

I am not familier with JOSE standards. It looks weird to me that I have to embed the certificate chain in `x5c` header for every JWT signed by auth server. Isn't that redundant of what is configured on the server side, via `rootcertbundle` option?

Instead of modifying auth server, I tried the new `jwks` option, and it worked for me. But it's somehow not easy to convert a EC P-256 PEM public key to a JWK, as many tools I looked for only support RSA. I didn't find easier way to retrieve `kid` generated by `libtrust` on the auth server side, other than looking into distribution error log. I have to append `kid` field to the converted JWK, and construct the JWKS structure manually. The overall task is not trivial.

Is it possible to provide backward compatibilty without requiring `x5c` header? Or is it possible to improve the documentation to guide people how to adopt the modified token verification process?

**Comment by milosgajdos** [comment]
> Is it possible to provide backward compatibilty without requiring x5c header? Or is it possible to improve the documentation to guide people how to adopt the modified token verification process?

I am curious why this is causing you issues when migrating from `v2.8` because the `x5c` check has always been there https://github.com/distribution/distribution/blob/5d5c60f77c01b143b22be9d62af3b32843fbcb0c/registry/auth/token/token.go#L186-L222

Updating the docs is always a great idea; do you fancy opening a PR?

**Comment by zhangyoufu** [comment]
@milosgajdos I didn't use `x5c` way to authenticate.

Before #4096 `feat: replace docker/libtrust with go-jose/go-jose`, the trusted JWKs always come from certificate chain, the `kid` is generated using `libtrust`, a non-standard(?) `keyIDFromCryptoKey`.

After #4096, we have to configure trusted jwks manually.

related code in v2.8

https://github.com/distribution/distribution/blob/5d5c60f77c01b143b22be9d62af3b32843fbcb0c/registry/auth/token/accesscontroller.go#L214-L223

related code in v3.0

https://github.com/distribution/distribution/blob/3fe707de5c94160f40b41454e0fd0b91921d0176/registry/auth/token/accesscontroller.go#L337-L347

I believe that many auth servers in use today rely on v2.8 behavior.

**Comment by milosgajdos** [comment]
Ah I see, yeah we've changed that. I am not sure if considered to return to the original logic. 

We saw v3 as an opportunity to ditch the now archived `libtrust` module and established more explicit code when configured trusted keys, etc.

**Comment by zhangyoufu** [comment]
A seperate configurable `jwks` option is nice to have, as it allows more keys to be trusted (maybe used in different geo-location) and rotated in their own pace. But please, don't break this just because we have an opportunity (the v3 release) to break it. We can extract the keyid generation logic without relying on `libtrust`.

And by the way, didn't your change make `rootcertbundle` useless? I found no consumer of `rootPool`.

**Comment by zhangyoufu** [comment]
> We can extract the keyid generation logic without relying on `libtrust`.

I proposed PR #4417. I understand that we didn't reach consensus on whether we should or should not break this, but PTAL.
